src/meta-srv/src/lease.rs

```diff
@@ -81,16 +118,77 @@ pub async fn lookup_datanode_peer(
     }
 }
 
+type LeaseFilterFuture<'a, K> =
+    Pin<Box<dyn Future<Output = Result<HashMap<K, LeaseValue>>> + Send + 'a>>;
+
+pub struct LeaseFilter<'a, K>
+where
+    K: Eq + Hash + TryFrom<Vec<u8>, Error = Error> + 'a,
+{
+    lease_secs: u64,
+    key_prefix: Vec<u8>,
+    meta_peer_client: &'a MetaPeerClientRef,
+    condition: Option<fn(&LeaseValue) -> bool>,
+    inner_future: Option<LeaseFilterFuture<'a, K>>,
+}
+
+impl<'a, K> LeaseFilter<'a, K>
+where
+    K: Eq + Hash + TryFrom<Vec<u8>, Error = Error> + 'a,
+{
+    pub fn new(
+        lease_secs: u64,
+        key_prefix: Vec<u8>,
+        meta_peer_client: &'a MetaPeerClientRef,
+    ) -> Self {
+        Self {
+            lease_secs,
+            key_prefix,
+            meta_peer_client,
+            condition: None,
+            inner_future: None,
+        }
+    }
+
+    /// Set the condition for the lease filter.
+    pub fn with_condition(mut self, condition: fn(&LeaseValue) -> bool) -> Self {
+        self.condition = Some(condition);
+        self
+    }
+}
+
+impl<'a, K> Future for LeaseFilter<'a, K>
+where
+    K: Eq + Hash + TryFrom<Vec<u8>, Error = Error> + 'a,
+{
+    type Output = Result<HashMap<K, LeaseValue>>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+
+        if this.inner_future.is_none() {
+            let lease_filter = build_lease_filter(this.lease_secs);
+            let condition = this.condition;
+            let fut = filter(this.key_prefix.clone(), this.meta_peer_client, move |v| {
+                lease_filter(v) && condition.unwrap_or(|_| true)(v)
+            });
```

Comment:
can we take to avoid clone here?
```suggestion
            let condition = this.condition.take().unwrap_or(|_| true);
            let key_prefix = std::mem::take(&mut this.key_prefix);
            let fut = filter(key_prefix, this.meta_peer_client, move |v| {
                lease_filter(v) && condition(v)
            });
```